internal/pkg/sylog/sylog_test.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+// +build sylog
+
+package sylog
+
+import (
+	"strings"
+	"testing"
+)
+
+// TestSylogSetLevel to ensure the correct integer is returned
+func TestSylogSetLevel(t *testing.T) {
+	var levelTests = []struct {
+		name  string
+		level int
+	}{
+		{"SetLevelFatal", int(fatal)},
+		{"SetLevelError", int(error)},
+		{"SetLevelWarn", int(warn)},
+		{"SetLevelLog", int(log)},
+		{"SetLevelInfo", int(info)},
+		{"SetLevelVerbose", int(verbose)},
+		{"SetLevelVerbose2", int(verbose2)},
+		{"SetLevelVerbose3", int(verbose3)},
+		{"SetLevelDebug", int(debug)},
+	}
+
+	for _, tt := range levelTests {
+		t.Run(tt.name, func(t *testing.T) {
+			SetLevel(tt.level)
+
+			l := GetLevel()
+
+			if l != tt.level {
+				t.Errorf("got %d, want %d", l, tt.level)
+			}
+		})
+	}
+}
+
+func TestSylogPrefix(t *testing.T) {
+
+	var logSuffix = "\x1b[0m "
+	var levelTests = []struct {
+		name  string
+		level messageLevel
+	}{
+		{"\x1b[31mFATAL", fatal},
+		{"\x1b[31mERROR", error},
+		{"\x1b[33mWARNING", warn},
+		{"\x1b[0mLOG", log},
+		{"\x1b[34mINFO", info},
+		{"\x1b[0mVERBOSE", verbose},
+		{"\x1b[0mVERBOSE", verbose2},
+		{"\x1b[0mVERBOSE", verbose3},
+		{"\x1b[0mDEBUG", debug},
+	}
+
+	// Test that prefix is colored
+	for _, tt := range levelTests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			SetLevel(int(tt.level))
+			levelPrefix := prefix(tt.level)
+
+			// Check that we start with the color prefix
+			if !strings.HasPrefix(levelPrefix, tt.name) {
+				t.Errorf("got prefix %s, want %s", levelPrefix, tt.name)
+			}
+
+			// The suffix should be consistently the "off" color string
+			if !strings.HasSuffix(levelPrefix, logSuffix) && tt.level != debug {
+				t.Errorf("%s does not end with %s", levelPrefix, logSuffix)
+			}
+		})
+	}
+}
+
+func TestSylogDisableColor(t *testing.T) {
+
+	var logSuffix = "\x1b[0m "
+	var levelTests = []struct {
+		name  string
+		level messageLevel
+	}{
+		{"FATAL", fatal},
+		{"ERROR", error},
+		{"WARNING", warn},
+		{"LOG", log},
+		{"INFO", info},
+		{"VERBOSE", verbose},
+		{"VERBOSE", verbose2},
+		{"VERBOSE", verbose3},
+		{"DEBUG", debug},
+	}
+
+	// Disable all color output, removing prefix and off suffix
+	DisableColor()
+
+	// Test that prefix is colored
+	for _, tt := range levelTests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			SetLevel(int(tt.level))
+			levelPrefix := prefix(tt.level)
+
+			// Check that we start with the color prefix
+			if !strings.HasPrefix(levelPrefix, tt.name) {
+				t.Errorf("got prefix %s, want %s", levelPrefix, tt.name)
+			}
+
+			// The suffix should be consistently the "off" color string
+			if strings.HasSuffix(levelPrefix, logSuffix) && tt.level != debug {
```

Comment:
I'm not sure why this isn't failing, if we have called `DisableColor`, then we should never be including any color codes in our return value from `prefix()`

Reply:
That means it's correct - the error is only called if the return value (levelPrefix) has the suffix (which it doesn't because it's disabled).

Reply:
That makes sense. Why do you need the extra tt.level != debug, though?

Reply:
Good question! When debug is the level, there is the added uid and process id to the output (something like `[uid=1000,pid=1234]`) and it's not colored, so it doesn't have the same suffix as the others.

Reply:
So, I think it's actually probably easier to just match a regex here instead of trying to handle every individual case. We can create a color regex, and a no-color regex:

To match for colored prefixes: `^(\\x1b\[\d+m)(\w{1,7}\s*)(\\x1b\[0m)`
To match for without prefixes: `^(FATAL|ERROR|WARNING|LOG|INFO|VERBOSE|DEBUG)\s*`

I think we can just apply those regexps on the returned prefixes. The first should only match for prefixes returned in `TestPrefix`, and the second one should match only for prefixes returned within `TestDisableColor`. I think this will greatly simplify the possibility that we've gotten something wrong in the testing matrix

Reply:
Actually that second one needs a small tweak, hold on

Reply:
OK, second regex is correct now I believe

Reply:
A single regex would be a lot cleaner, but Isn't the idea of matching the messageLevels to strings to check for that? For example, we want to check that "info" (the integer / messageLevel) corresponds to the string "INFO." If we do one gigantic regular expression, we don't check for that.